ssl/s3_lib.c

```diff
@@ -4148,6 +4151,32 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
     if (s->options & SSL_OP_CIPHER_SERVER_PREFERENCE || tls1_suiteb(s)) {
         prio = srvr;
         allow = clnt;
+#ifndef OPENSSL_NO_CHACHA
+        /* If ChaCha is at the top of the client preference list,
+           temporarily prioritize ChaCha ciphers in the servers list. */
+        if (s->options & SSL_OP_PRIORITIZE_CHACHA && sk_SSL_CIPHER_num(clnt) > 0) {
+            c = sk_SSL_CIPHER_value(clnt, 0);
+            if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                /* Is client preferred, reprioritize... */
+                new_chacha = sk_SSL_CIPHER_new_null();
+                if (new_chacha != NULL) {
+                    /* Put ChaCha at the top */
+                    for (i = 0; i < sk_SSL_CIPHER_num(srvr); i++) {
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc == SSL_CHACHA20POLY1305)
+                            sk_SSL_CIPHER_push(new_chacha, c);
+                    }
+                    /* Pull in the rest */
+                    for (i = 0; i < sk_SSL_CIPHER_num(srvr); i++) {
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc != SSL_CHACHA20POLY1305)
+                            sk_SSL_CIPHER_push(new_chacha, c);
+                    }
```

Comment:
One can actually do it in single loop: `if (c->algorithm_enc == SSL_CHACHA20POLY1305) sl_SSL_CIPHER_unsift(new_chacha, c); else sk_SSL_CIPHER_push(new_chacha, c);`. I also wonder if compiler keeps calling sk_SSL_CIPHER_num for each iteration. I mean maybe it makes sense to ensure that it's called only once...

Reply:
Yes, it calls sk_SSL_CIPHER_num for each iteration.

Reply:
So, how would one prevent it from being called?

Reply:
Obviously
```
int num = sk_SSL_CIPHER_num(x)
for (i = 0; i < num; i++) {}
```